#! /usr/bin/env python

PACKAGE = 'ladybug_camera'

from dynamic_reconfigure.parameter_generator_catkin import *


class SensorLevels:
    RECONFIGURE_RUNNING = 0
    RECONFIGURE_STOP = 1


gen = ParameterGenerator()

mode_enum = gen.enum([gen.const("Off", int_t, 0, "Turn off configuration"),
                      gen.const("Manual", int_t, 1, "Manual configuration"),
                      gen.const("Auto", int_t, 2, "Automatic configuration")],
                     "An enum to set size")

#       Name                                Type        Reconfiguration level             Description                                                                                              Default                     Min      Max
gen.add("frame_rate", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "User controlled acquisition frame rate in Hertz (frames per second).", 10, 0, 120)
gen.add("frame_rate_enable", bool_t, SensorLevels.RECONFIGURE_RUNNING,
        "Enables manual (true) and automatic (false) control of the aquisition frame rate", False)

# Gain Settings
# Control the amount of amplification applied to a pixel by the camera A/D converter
gen.add("gain_state", int_t, 0, "Gain mode - Auto, Manual, Off", 2, 0, 2, edit_method=mode_enum)
gen.add("gain", double_t, SensorLevels.RECONFIGURE_RUNNING, "Set the gain in dB when in manual mode (0-18dB)", 2, 0,
        180)

# Set Exposure
# Note: For the Auto Exposure 'auto' feature, gain and/or exposure time must be set to auto.                               "Timed")
gen.add("exposure_state", int_t, 0, "Sets the automatic exposure mode to one of the states: 'Off', 'Manual', 'Auto'", 2,
        0, 2, edit_method=mode_enum)
gen.add("exposure", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Sets an exposure value (EV) when Auto Exposure is in the Manual Mode", -0.85, -7.585, 2.414)

# Shutter Settings
gen.add("shutter_state", int_t, 0, "Shutter mode - Auto, Manual, Off", 2, 0, 2, edit_method=mode_enum)
gen.add("shutter", double_t, SensorLevels.RECONFIGURE_RUNNING, "Set the shutter time in ms when in manual mode (2-?)",
        9.83191, 0.0268817, 34.102)

# Gamma
# Turning gamma off removes gamma correction
gen.add("gamma_state", bool_t, SensorLevels.RECONFIGURE_RUNNING, "On / Off to control gamma correction.", True)
gen.add("gamma", double_t, SensorLevels.RECONFIGURE_RUNNING, "Controls the gamma correction of pixel intensity.", 2.2,
        0.5, 4.0)

# White Balance
gen.add("white_balance_state", int_t, 0, "Configure running mode for white balance", 2, 0, 2, edit_method=mode_enum)
gen.add("white_balance_blue_ratio", double_t, SensorLevels.RECONFIGURE_RUNNING, "White balance blue component.", 800, 0,
        1023)
gen.add("white_balance_red_ratio", double_t, SensorLevels.RECONFIGURE_RUNNING, "White balance red component.", 550, 0,
        1023)

# Brightness = Black Level
gen.add("brightness", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Also known as Black level offset. Refers to the output of the camera when not illuminated.", 1.5, 0.0, 12.4756)

exit(gen.generate(PACKAGE, "ladybug_camera", "Ladybug"))
